src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -874,46 +872,106 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+
+              if (doEtaZero && settings.warnEtaZero) {
+                val ptHelp =
+                  if (expectingFunctionOfArity) pt
+                  else s"$pt, which is SAM-equivalent to ${samToFunctionType(pt)}"
+
+                reporter.warning(tree.pos, s"An unapplied 0-arity method was eta-expanded (due to the expected type ${ptHelp}), rather than applied to `()`.\n" +
+                                           s"Write ${Apply(warnTree, Nil)} to invoke method ${meth.decodedName}, or change the expected type.")
+              }
+              doEtaZero
+            } else sourceLevel2_14 || expectingFunctionOfArity || expectingSamOfArity
+
+          if (doIt && !expectingFunctionOfArity) warnEtaSam()
+
+          doIt
         }
-        // (4.3) eta-expand method value when function or sam type is expected (for experimentation, always eta-expand under 2.14 source level)
-        else if (isFunctionProto(pt) || settings.isScala214) { // TODO: decide on `settings.isScala214`
-          if (settings.isScala212 && mt.params.isEmpty) // implies isFunctionType(pt)
-             currentRun.reporting.deprecationWarning(tree.pos, NoSymbol, "Eta-expansion of zero-argument methods is deprecated. "+
-                    s"To avoid this warning, write ${Function(Nil, Apply(tree, Nil))}.", "2.12.0")
 
-          typedEtaExpansion(tree, mode, pt)
+
+        // (4.2) condition for auto-application by -Xsource level
+        //
+        // until 2.14: none (assuming condition for (4.3) was not met)
+        // in 3.0: `meth.isJavaDefined`
+        //         (TODO decide -- currently the condition is more involved to give slack to Scala methods overriding Java-defined ones;
+        //          I think we should resolve that by introducing slack in overriding e.g. a Java-defined `def toString()` by a Scala-defined `def toString`.
+        //          This also works better for dealing with accessors overriding Java-defined methods. The current strategy in methodSig is problematic:
+        //          > // Add a () parameter section if this overrides some method with () parameters
+        //          > val vparamSymssOrEmptyParamsFromOverride =
+        //          This means an accessor that overrides a Java-defined method gets a MethodType instead of a NullaryMethodType, which breaks lots of assumptions about accessors)
+        def checkCanAutoApply(): Boolean = {
+          if (sourceLevel2_14 && !meth.isJavaDefined)
+            context.deprecationWarning(tree.pos, NoSymbol, s"Auto-application to `()` is deprecated. Write ${Apply(warnTree, Nil)} to invoke method ${meth.decodedName}," +
+                                                           s"or remove the empty argument list from its definition (Java-defined methods are exempt).\n"+
+                                                           s"In Scala 3, an unapplied method will be eta-expanded, as in: ${Function(Nil, Apply(warnTree, Nil))}.", "2.14.0")
```

Comment:
I'm a little skittish about echoing `tree.toString` in compiler error messages. The intent is to make things copy/pastable into source code, but if the tree is large (maybe it contains macro expansions or other desugarings) the message itself is hard to see.

Reply:
A compiler option to copy advice directly to the clipboard! or `reporters.Clippy`.

Reply:
Dropped tree toString, added more advice to compensate.